src/core/lib/channel/http_client_filter.c

```diff
@@ -134,17 +150,57 @@ static void hc_mutate_op(grpc_call_element *elem,
   /* grab pointers to our data from the call element */
   call_data *calld = elem->call_data;
   channel_data *channeld = elem->channel_data;
+
+  /* Decide which HTTP VERB to use */
+  grpc_mdelem *method = GRPC_MDELEM_METHOD_POST;
+  if (op->send_initial_metadata != NULL && (op->send_initial_metadata_flags &
+       GRPC_INITIAL_METADATA_CACHEABLE_REQUEST) &&
+      op->send_message != NULL &&
+      op->send_message->length < channeld->max_payload_size_for_get) {
+    method = GRPC_MDELEM_METHOD_GET;
+  } else if (op->send_initial_metadata_flags &
+             GRPC_INITIAL_METADATA_IDEMPOTENT_REQUEST) {
+    method = GRPC_MDELEM_METHOD_PUT;
+  }
+
+  if (method == GRPC_MDELEM_METHOD_GET) {
+    gpr_slice slice;
+    gpr_slice_buffer slices;
+    gpr_slice_buffer_init(&slices);
+    /* allocate memory to hold the entire payload */
+    calld->payload_bytes = gpr_malloc(op->send_message->length);
+    GPR_ASSERT(calld->payload_bytes);
+    uint8_t *wrptr = calld->payload_bytes;
+
+    /* copy payload from slices into payload_bytes. It gets freed in
+     * op_complete*/
+    while (grpc_byte_stream_next(NULL, op->send_message, &slice, ~(size_t)0,
```

Comment:
Bad usage of grpc_byte_stream_next.

```
/* returns 1 if the bytes are available immediately (in which case
 * on_complete will not be called), 0 if the bytes will be available
 * asynchronously.
 *
 * max_size_hint can be set as a hint as to the maximum number
 * of bytes that would be acceptable to read.
 *
 * once a slice is returned into *slice, it is owned by the caller.
 */
int grpc_byte_stream_next(grpc_exec_ctx *exec_ctx,
                          grpc_byte_stream *byte_stream, gpr_slice *slice,
                          size_t max_size_hint, grpc_closure *on_complete);
```

This needs a closure chain to allow asynchronous fetching of the bytes. Not doing so leaves us with a time-bomb.